broadinstitute/single_cell_portal_core
# Upload wizard visualization status indicators (SCP-3969)

SCP-3969  This adds indicators for whether a study has visualizations enabled for clusters and/or expression data.  It also polls the server every 10 seconds so that the indicators can be updated without need page refreshes, and spinners can be shown for when a file is parsing.  

The approach for this polling was to keep it as simple as possible--every 10 seconds it asks the server to resend all information about the study and study files.  This avoids needing complex logic of inter-file dependencies in determining what to poll.  Once the client receives the updated study data, it stores it in the 'serverState' variable, and it also matches each study form file object to the corresponding server file state.  The server state is pointed to in a 'serverFile' property on each form file.  So in rough [MVVM ](https://www.educative.io/collection/page/5429798910296064/5725579815944192/5161641743220736 )terms, the serverState/serverFile is the model, the formState/file is the view model, and the components are the view.  

Two new spinners indicate when a file is awaiting remote, or parsing:  
![image (9)](https://user-images.githubusercontent.com/2800795/148494925-528f26c6-d0a9-4f91-9b9b-113f16dd1b3f.png)

![image (7)](https://user-images.githubusercontent.com/2800795/148494897-9c2bd72e-6f03-402a-8305-f9d78c89fefd.png)

Visual indicators show that the study is missing files needed for visualization
![image (8)](https://user-images.githubusercontent.com/2800795/148494990-57a2316b-1417-4122-8314-f73d06125ed8.png)

TO TEST:
1. Open an existing study with visualizations (e.g. male mouse brain), in the upload wizard, confirm the indicators are green
2. Create a new study -- confirm the indicators show no visualizations yet
3. add expression matrices.  Confirm the spinners behave as expected
4. add a metadata file, confirm spinners behave as expected during loading, and once parsing completes, 'expression visuals' indicator should be green
5. add a cluster file, confirm 'cluster visuals' indicator is green once parsing completes.
